File: androidFood2Fork/src/main/java/com/egorshustov/food2forkkmm/android/presentation/recipe_detail/RecipeDetailViewModel.kt
package com.egorshustov.food2forkkmm.android.presentation.recipe_detail

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.egorshustov.food2forkkmm.android.util.RECIPE_ID_ARG
import com.egorshustov.food2forkkmm.domain.util.Result
import com.egorshustov.food2forkkmm.presentation.model.GenericMessageInfo
import com.egorshustov.food2forkkmm.presentation.model.UIComponentType
import com.egorshustov.food2forkkmm.presentation.recipe_detail.RecipeDetailEvent
import com.egorshustov.food2forkkmm.presentation.recipe_detail.RecipeDetailState
import com.egorshustov.food2forkkmm.usecases.recipe_detail.GetRecipeUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import java.util.UUID
import javax.inject.Inject
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

@HiltViewModel
class RecipeDetailViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val getRecipeUseCase: GetRecipeUseCase
) : ViewModel() {

    val state: MutableState<RecipeDetailState> = mutableStateOf(RecipeDetailState())

    init {
        savedStateHandle.get<Int>(RECIPE_ID_ARG)?.let { recipeId ->
            onTriggerEvent(RecipeDetailEvent.GetRecipe(recipeId))
        }
    }

    fun onTriggerEvent(event: RecipeDetailEvent) {
        when (event) {
            is RecipeDetailEvent.GetRecipe -> getRecipe(event.recipeId)
            else -> appendToMessageQueue(
                GenericMessageInfo.Builder()
                    .id(UUID.randomUUID().toString())
                    .title("Error")
                    .uiComponentType(UIComponentType.Dialog)
                    .description("Unknown Event")
            )
        }
    }

    private fun getRecipe(id: Int) = with(state.value) {
        getRecipeUseCase(id).onEach { result ->
            state.value = when (result) {
                is Result.Success -> copy(recipe = result.data, isLoading = false)

                is Result.Error -> {
                    appendToMessageQueue(
                        GenericMessageInfo.Builder()
                            .id("GetRecipe.Error")
                            .title("Error")
                            .uiComponentType(UIComponentType.Dialog)
                            .description(result.exception.message ?: "Unknown Error")
                    )
                    copy(isLoading = false)
                }

                Result.Loading -> copy(isLoading = true)
            }
        }.launchIn(viewModelScope)
    }

    private fun appendToMessageQueue(messageInfoBuilder: GenericMessageInfo.Builder) {
        val queue = state.value.queue
        queue.add(messageInfoBuilder.build())
        state.value = state.value.copy(queue = queue)
    }
}

File: iosFood2Fork/iosFood2Fork/presentation/recipe_detail/RecipeDetailViewModel.swift
//
//  RecipeDetailViewModel.swift
//  iosFood2Fork
//
//  Created by Egor on 14.12.2021.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation

import SwiftUI
import shared

class RecipeDetailViewModel: ObservableObject {
    
    // Dependencies
    private let getRecipeUseCase: GetRecipeUseCase
    
    @Published var state: RecipeDetailState = RecipeDetailState()
    
    init(
        recipeId: Int,
        getRecipeUseCase: GetRecipeUseCase
    ) {
        self.getRecipeUseCase = getRecipeUseCase
        onTriggerEvent(event: RecipeDetailEvent.GetRecipe(recipeId: Int32(recipeId)))
    }
    
    func onTriggerEvent(event: RecipeDetailEvent) {
        switch event {
            case is RecipeDetailEvent.GetRecipe:
                getRecipe(recipeId: Int((event as! RecipeDetailEvent.GetRecipe).recipeId))
            case is RecipeDetailEvent.OnRemoveHeadMessageFromQueue:
                removeHeadMessageFromQueue()
            default:
                self.appendToMessageQueue(messageInfoBuilder: GenericMessageInfo.Builder()
                    .id(id: UUID().uuidString)
                    .title(title: "Error")
                    .uiComponentType(uiComponentType: UIComponentType.Dialog())
                    .description(description: "Unknown Event")
                    .positive(positiveAction: PositiveAction(positiveBtnTxt: "OK", onPositiveAction: {})))
        }
    }
    
    private func getRecipe(recipeId: Int){
        getRecipeUseCase.invoke(
            id: Int32(recipeId)
        ).collectCommon(
            coroutineScope: nil,
            callback: { result in
                switch result {
                        case is ResultSuccess<AnyObject>: do {
                            let resultSuccess = result as! ResultSuccess<AnyObject>
                            let recipe = resultSuccess.data as! Recipe
                            self.updateState(isLoading: false, recipe: recipe)
                        }
                        case is Result.Error: do {
                            let resultError = result as! Result.Error
                            let exception = resultError.exception
                            print(exception)
                            self.appendToMessageQueue(messageInfoBuilder: GenericMessageInfo.Builder()
                                .id(id: "GetRecipe.Error")
                                .title(title: "Error")
                                .uiComponentType(uiComponentType: UIComponentType.Dialog())
                                .description(description: exception.message ?? "Unknown Error")
                                .positive(positiveAction: PositiveAction(positiveBtnTxt: "OK", onPositiveAction: {})))
                            self.updateState(isLoading: false)
                        }
                        case is Result.Loading: do {
                            print("Loading")
                            self.updateState(isLoading: true)
                        }
                        default: do {
                            print("default")
                        }
                }
            }
        )
    }
    
    private func updateState(
            isLoading: Bool? = nil,
            recipe: Recipe? = nil,
            queue: Queue<GenericMessageInfo>? = nil
    ){
        let currentState = (self.state.copy() as! RecipeDetailState)
        self.state = self.state.doCopy(
            isLoading: isLoading ?? currentState.isLoading,
            recipe: recipe ?? currentState.recipe,
            queue: queue ?? currentState.queue
        )
    }
    
    private func appendToMessageQueue(messageInfoBuilder: GenericMessageInfo.Builder) {
        let currentState = self.state.copy() as! RecipeDetailState
        let queue = currentState.queue
        let queueUtil = GenericMessageInfoQueueUtil()
        let messageInfo = messageInfoBuilder.build()
        if !queueUtil.doesMessageAlreadyExistInQueue(queue: queue, messageInfo: messageInfo) {
            queue.add(element: messageInfo)
            updateState(queue: queue)
        }
    }
    
    private func removeHeadMessageFromQueue() {
        let currentState = self.state.copy() as! RecipeDetailState
        let queue = currentState.queue
        do {
            try queue.remove()
            updateState(queue: queue)
        }
        catch {
            print("\(error)")
        }
    }
    
}
